### DISCUSSION TITLE
What the status of the `?` and `!` Option/Result return types?

### DISCUSSION THREAD

**Discussion by srackham** [vlang/v#16143] [General]
On 11-Jul-2022 [@medvednikov commented](https://github.com/vlang/v/issues/7674#issuecomment-1179773418):

> Option and Result have been split up: ?Foo and !Foo.

This seems to contradicted by the [current documentation](https://github.com/vlang/v/blob/master/doc/docs.md#optionresult-types-and-error-handling), which I'm guessing is out of date:

> V combines Option and Result into one type, so you don't need to decide which one to use.

The same documentation also states:

> If you don't need to return an error message, you can simply return none (this is a more efficient equivalent of return error("")).

There is also this RFC, updated 21-Apr-2022: [Redesign of error and none handling](https://github.com/vlang/rfcs/blob/master/error_handling.md) but it's implementation status is unclear (to me).

From what I can tell things as they currently stand can be sumarised as follows:

- `!T` returns a `Result` type and should be used if you want to return a value or an error?
- `?T` returns an `Option` type and should be used if you want to return a value or `none`?
- 3-way return types (value, `none` and error), which are currently allowed in `?` functions, are deprecated?

See also #14664

**Comment by JalonSolov** [comment]
The status is... in transition.

`?Type` still works as it used to, but the changes are being made to separate `?Type` and `!Type`.

The documentation will be updated when `?Type` no longer works as it did in the past.

**Comment by srackham** [comment]
@JalonSolov Thanks. I'm assuming my summary was correct and from now on should code accordingly.

**Reply by JalonSolov** [reply]
Yes.  The new way works already.
